### ISSUE TITLE
Consider passing Databases on the constructor for the classes which need them

### ISSUE THREAD

**Issue by seanrobbins** [tronprotocol/java-tron#35]
UTXOStore and UTXOSet classes both initialize and set a lock on the TransactionDB.

Blockchain and BlockStores classes both initialize and set a lock on the BlockDB.

In all cases these initialization on the DB are done in the constructor of the class.

It seems like these DBs could be initialized outside of the class, perhaps in the Peer class and then the database dependency injected into Blockchain, BlockStores, UTXOStore and UTXOSet in the constructor.

This way, we can ensure the DB connections are created and locked only once during the runtime lifecycle and avoid the NativeDB$DBException.

This would also allow using a mock for TransactionDB and BlockDB, which would allow for easier state management of the data in the unit tests and so better test coverage.

A wrapper class could be created around the initialization of the database so that the type system can make sure the correct DB is injected into correct class. For example, the Blockchain constructor could look like:

`public Blockchain(BlockDB blockDb)`

Any thoughts on this approach?

**Comment by sasaxie** [comment]
Or unit test uses the test database directory (database-test) and Tron uses the official database directory(database-official).

**Comment by Rovak** [comment]
I made a start in the PR #45

**Comment by seanrobbins** [comment]
BlockStores class is still initializing its own connection to the BlockDB, which might cause DB lock exceptions in the future.

**Comment by seanrobbins** [comment]
I'm going to close this issue for now. BlockStores is initializing its own DB connection: https://github.com/tronprotocol/java-tron/blob/7d6cab704c9081b96dd862579728fe752f0d9d9c/src/main/java/org/tron/dbStore/BlockStores.java#L28

I'm not sure if this may cause DB lock issue in the future, I'll look into it a bit more and either submit a PR with refactor or open a new ticket.
